rocSPARSE changes to support multiple ROCM installation

pruthvistony opened the pull request:
- ROCm stack install path needs to be provided using the
  variable ROCM_PATH as a cmake parameter. If it is not
  provided, ROCm stack components are searched in /opt/rocm
- Package is generated to install into ROCM_PATH by setting
  it to CMAKE_PACKAGING_INSTALL_PREFIX, if defined in the env
  otherwise default into /opt/rocm
- RUNPATH is set in the lib only if ROCM_RPATH env is defined
  with the rpath


Signed-off-by: Pruthvi Madugundu <mpruthvi@gmail.com>

ntrost57 commented:
I am not able to generate packages (make package) for CentOS with this changes:
```CPack3: Create package
CMake Error at /usr/share/cmake3/Modules/Internal/CPack/CPackRPM.cmake:125 (file):
  file RELATIVE_PATH must be passed a full path to the directory:
  rocsparse-install```

saadrahim commented on library/CMakeLists.txt:
```diff
@@ -124,7 +129,7 @@ set(CPACK_RPM_EXCLUDE_FROM_AUTO_FILELIST_ADDITION "\${CPACK_PACKAGING_INSTALL_PR
 set(package_name rocsparse)
 
-set(ROCSPARSE_CONFIG_DIR "\${CPACK_PACKAGING_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}" CACHE PATH "Path placed into ldconfig file")
+set(ROCSPARSE_CONFIG_DIR "/opt/rocm/${CMAKE_INSTALL_LIBDIR}" CACHE PATH "Path placed into ldconfig file")
```
I do not agree that /opt/rocm/ should be hard coded.

pruthvistony replied:
This is kept /opt/rocm as we want the normal users to use the ldconfig path to find the dependent libs with a symlink(/opt/rocm -> /opt/rocm-ver) and advanced users will use the RUNPATH to find all the dependent libs

saadrahim commented on cmake/Dependencies.cmake:
```diff
@@ -31,7 +31,7 @@ find_package(rocprim REQUIRED)
 # Workaround until hcc & hip cmake modules fixes symlink logic in their config files.
 # (Thanks to rocBLAS devs for finding workaround for this problem!)
-list(APPEND CMAKE_PREFIX_PATH /opt/rocm/hcc /opt/rocm/hip /opt/rocm)
+list(APPEND CMAKE_PREFIX_PATH ${CMAKE_INSTALL_PREFIX}/hcc ${CMAKE_INSTALL_PREFIX}/hip ${CMAKE_INSTALL_PREFIX})
```
CMAKE_INSTALL_PREFIX shouldn't be used for library dependency searches. A user can always desire to install in a different location such a home directory.

pruthvistony commented:
> 
> 
> I am not able to generate packages (make package) for CentOS with this changes:
> 
> ```
> CMake Error at /usr/share/cmake3/Modules/Internal/CPack/CPackRPM.cmake:125 (file):
>   file RELATIVE_PATH must be passed a full path to the directory:
>   rocsparse-install
> ```
Checking on the failure on Centos.

pruthvistony commented:
Updated the changes and removed the usage of CMAKE_INSTALL_PREFIX to search for library path.
Please check the latest commit. Need to check the CI failures after this commit.
